### DISCUSSION TITLE
Eclipse Layout Kernel (ELK) integrtion?

### DISCUSSION THREAD

**Discussion by PatrikN-work** [eclipse-glsp/glsp#1027] [Q&A]
How do I integrate and configure ELK for the macro layout for a diagram?

**Comment by martin-fleck-at** [comment]
Hi @PatrikN-work, unfortunately I am not too familiar with the ELK layout engine but hopefully @CamilleLetavernier can help you out here.

**Reply by PatrikN-work** [reply]
Hi @CamilleLetavernier  - sorry for reviving this, but do you have any good description of how to integrate ELk in GLSP in Node/TS, not java?

**Reply by CamilleLetavernier** [reply]
Hi Patrik,

I will be of limited help with this, as I've never used ELK with GLSP, and in general my experience with ELK is very limited. I'll try to give a few general hints nonetheless.

I don't think you can use ELK with Node/TS directly. AFAIK, it only exists as a Java plug-in; so you would need to trigger it on the Java side. Since the Java server doesn't know about the client-side layout (especially the rendered size of SVG elements), it needs to happen in several steps:

- the java server sends the diagram with minimal layout information (e.g. position of elements, and maybe size hints, but usually no actual size)
- the diagram is rendered in SVG on the client side. This determines the actual rendered size of elements (which may be affected by things like SVG Renderers, CSS Fonts, Border widths, etc. that the java server can't take into account)
- the client sends a ComputedBounds action to the server with this layout information. Now the server has an idea about the actual rendered size of elements.
- the server can now trigger the ELK Layout on these elements and send the new positions (and maybe size constraints) to the client

Note that it **might** be possible to avoid the client-server round-trip if you expect to be able to "force" sizes on the server directly (i.e. if your SVG Renderers/CSS are not expected to give dynamic sizes to your elements. For example, if you use text-ellipsis for Node labels, then the font size won't force the nodes to grow bigger than expected). In that case, you can trigger the ELK Layout directly in the first step, before sending the diagram to the client.

There are a few options and actions involved in the way the layout is computed (Client-only, server-only or round-trip):

- ViewerOptions.needsClientLayout: true if you expect a roundtrip (or client-only layouting).
- ComputedBoundsAction: action sent by the client to the server after client-side layout has been applied. This informs the server about the actual "rendered size" of elements

Beside my lack of experience with ELK, there are many variables that make it difficult to precisely answer this question. Most importantly:

- GLSP now includes many different options for the backend. I believe ELK only works if you use a GLSP Java Server (with or without a Model Server).
- There is a lot of freedom in the way you render and layout your nodes on the client (SVG+CSS). Depending on which features you use, you may be able to do server-side layout, or might need to do a full round-trip.
- The ELK layout itself can influence the size of nodes, or their position, or both. If you only care about positioning, you might not need the full roundt-rip. If sizes need to be taken into account, or forced to a specific value by the ELK Layout, then the round-trip may or may not be necessary.

I hope this helps!

**Reply by planger** [reply]
One small addition: the node-based GLSP server uses [`elkjs`](https://github.com/kieler/elkjs). So it isn't necessary to talk to a Java server. You can do it, which I believe may have advantages with respect to performance, but you don't need to.

**Reply by PatrikN-work** [reply]
Hi Camille,

Thanks for the answer, it sums up my (confused) understanding that it seems complex, especillay the double roundtrip.

/Patrik

PS. i actually found this now, not sure why I didnät stumble on it before: https://github.com/kieler/elkjs DS

**Reply by PatrikN-work** [reply]
"the node-based GLSP server uses [elkjs](https://github.com/kieler/elkjs)"  

But how do I invoke it. Just like this?

`elk.layout(graph)
`
on the Gmodel?

**Reply by planger** [reply]
You need to trigger a `LayoutAction`, this should end up calling the [ElkLayoutEngine](https://github.com/eclipse-glsp/glsp-server-node/blob/main/packages/layout-elk/src/glsp-elk-layout-engine.ts). However, it only operates on the GModel, not your source model. So you would need to apply the computed positions back to your source model yourself. See also https://github.com/eclipse-glsp/glsp/issues/1173

**Reply by PatrikN-work** [reply]
That is if I want to persist the layout right? 

Doing a layout n the gmodel will trigger it to be sent to the client? 

Or do I need to do that myself?

**Reply by PatrikN-work** [reply]
In tried with this (Didn't find any LayoutAction, I guess it is the LayoutOperation):

`        containerElement.appendChild(this.createButton('btn_center_diagram', 'layout', 'screen-full', LayoutOperation.create([])));`

Hovever the only thing happening is that the diagram flashes (I suppose it is redrawn) but no layout chnages.  Is this since I don't "apply the computed positions back to your source model "?

**Reply by planger** [reply]
Oh, I'm sorry. Yes it is called `LayoutOperation` (it has been renamed with 2.x).

The `ElkLayoutEngine` just applies this to the GModel of your `modelState`. So if you create your GModel from another source model, then, without knowing your code, I assume you are right, it is because the changes aren't written back from the GModel to the source, but the model update causes the GModel to be overwritten again from your source model.

To resolve https://github.com/eclipse-glsp/glsp/issues/1173, we are thinking of deriving a set of `ChangeBoundsOperations` from the changed diagram after the auto-layout and apply them the source model via the handlers a GLSP adopter would have to provide anyway to react to user move operations. Then this problem would be solved.
However, right now you'd need to do that yourself.

**Reply by PatrikN-work** [reply]
Si get this in the log when calling layout fom the client.

![image](https://github.com/eclipse-glsp/glsp/assets/119344948/6bf4a07b-bcc6-40e7-bfc3-66177af33848)

Wher should I do this?

**Reply by planger** [reply]
There are a few options, but I think the easiest to workaround #1173 would be to overwrite and rebind the `LayoutOperationHandler` in your diagram server module. Something like
```ts
protected configureOperationHandlers(binding: InstanceMultiBinding<OperationHandlerConstructor>): void {
        super.configureOPerationHandlers(bidning);
        binding.rebind(LayoutOperationHandler).to(YourLayoutOperationHandler);
}
```

And in `YourLayoutOperationHandler` you could overwrite the method [`executeOperation`](https://github.com/eclipse-glsp/glsp-server-node/blob/4e5a7a9d306c282f5566497be5386a022f984afb/packages/server/src/common/features/layout/layout-operation-handler.ts#L59) invoke super and then walk through all GModel elements and create a `ChangeBoundsOperation`.

As a disclaimer: I haven't tried this out, that's just what comes to my mind. Maybe @tortmayr has a better idea.

**Reply by PatrikN-work** [reply]
Tried this:

        binding.rebind(LayoutOperationHandler, MyLayoutOperationHandler);

But MyLayoutOperationHandler.executeOperation() is never called

I try to triger it using:

        containerElement.appendChild(this.createButton('btn_center_diagram', 'layout', 'screen-full', LayoutOperation.create([])));

**Reply by PatrikN-work** [reply]
ok. execute() is triggered, however the list of elements in the layout operation is epty:

    Client (oat-diagram_0): layout {kind: 'layout', isOperation: true, elementIds: Array(0)} which according ot the docs means "default to the root **element". 
    
I just don't get how to get the layouted GModel....

**Reply by PatrikN-work** [reply]
Hmm. Do I need to do soemthing like this from the sprotty-nested-demo:

![image](https://github.com/eclipse-glsp/glsp/assets/119344948/79cc7408-6747-463c-8ede-90caf3613f6a)

**Reply by planger** [reply]
To configure the layout engine in GLSP, you'll need a `LayoutConfigurator`, e.g. in the workflow example:
https://github.com/eclipse-glsp/glsp-server-node/blob/main/examples/workflow-server/src/common/layout/workflow-layout-configurator.ts

And its registration in the module:
https://github.com/eclipse-glsp/glsp-server-node/blob/main/examples/workflow-server/src/browser/app.ts#L30

**Reply by PatrikN-work** [reply]
Thanks @planger, one step forward. 

Now I do get access to the Gmodel but I'm not sure how to ('walk through all GModel elements) and create a ChangeBoundsOperation.'


Regarding Operation, any pkan to update this? https://eclipse.dev/glsp/documentation/modeloperations/
How obsolete is it?

**Reply by PatrikN-work** [reply]
Also, now I get this for some reason 

    Server (oat-diagram_0): message {kind: 'message', message: 'No constructor is configured in DiagramConfiguration for type node1', severity: 'ERROR'}

**Reply by planger** [reply]
To apply the changes of the layout engine to your source model, as mentioned above, I'd probably go with overwriting the LayoutOperationHandler and dispatch the ChangeBoundsOperations. But I'd need to try this myself to be sure what's the best approach. We should probably just make sure to address https://github.com/eclipse-glsp/glsp/issues/1173 soon to avoid such a workaround.

**Comment by soerendomroes** [comment]
Hi @PatrikN-work, you would want to call ELK in plain java as described [here](https://eclipse.dev/elk/documentation/tooldevelopers/usingplainjavalayout.html). If you have questions on how to configure ELK feel free to ask in its [gitter chat](https://app.gitter.im/#/room/#eclipse_elk:gitter.im).
